File: ruby/lib/solana_mpp/intents/subscription_request.rb
# frozen_string_literal: true

module SolanaMpp
  module Intents
    class SubscriptionRequest
      PERIOD_DAY = 'day'
      PERIOD_WEEK = 'week'
      PERIOD_MONTH = 'month'

      attr_reader :amount,
                  :currency,
                  :period_unit,
                  :period_count,
                  :recipient,
                  :subscription_expires,
                  :description,
                  :external_id,
                  :method_details

      def initialize(amount:, currency:, period_unit:, period_count:, recipient: nil, subscription_expires: nil,
                     description: nil, external_id: nil, method_details: nil)
        ChargeRequest.assert_base_units(amount, 'amount')
        raise ArgumentError, 'currency is required' if currency.to_s.empty?
        @period_unit = self.class.normalize_period_unit(period_unit)
        ChargeRequest.assert_base_units(period_count, 'periodCount')
        raise ArgumentError, 'methodDetails must be a Hash' unless method_details.nil? || method_details.is_a?(Hash)

        @amount = amount
        @currency = currency
        @period_count = period_count
        @recipient = recipient
        @subscription_expires = subscription_expires
        @description = description
        @external_id = external_id
        @method_details = method_details
      end

      def to_h
        value = {
          'amount' => amount,
          'currency' => currency,
          'periodUnit' => period_unit,
          'periodCount' => period_count
        }
        value['recipient'] = recipient unless blank?(recipient)
        value['subscriptionExpires'] = subscription_expires unless blank?(subscription_expires)
        value['description'] = description unless blank?(description)
        value['externalId'] = external_id unless blank?(external_id)
        value['methodDetails'] = method_details unless method_details.nil?
        value
      end

      def self.from_h(value)
        raise ArgumentError, 'subscription request must be a Hash' unless value.is_a?(Hash)

        new(
          amount: value.fetch('amount', '').to_s,
          currency: value.fetch('currency', '').to_s,
          period_unit: value.fetch('periodUnit', '').to_s,
          period_count: value.fetch('periodCount', '').to_s,
          recipient: ChargeRequest.optional_string(value['recipient']),
          subscription_expires: ChargeRequest.optional_string(value['subscriptionExpires']),
          description: ChargeRequest.optional_string(value['description']),
          external_id: ChargeRequest.optional_string(value['externalId']),
          method_details: value['methodDetails']
        )
      end

      def self.normalize_period_unit(period_unit)
        return period_unit if [PERIOD_DAY, PERIOD_WEEK, PERIOD_MONTH].include?(period_unit)

        raise ArgumentError, "unsupported periodUnit: #{period_unit}"
      end

      private

      def blank?(value)
        value.nil? || value == ''
      end
    end
  end
end

File: php/src/Intent/SubscriptionRequest.php
<?php

declare(strict_types=1);

namespace SolanaMpp\Intent;

use DateTimeImmutable;
use InvalidArgumentException;

final class SubscriptionRequest
{
    public const PERIOD_DAY = 'day';
    public const PERIOD_WEEK = 'week';
    public const PERIOD_MONTH = 'month';

    /**
     * @param array<string, mixed>|null $methodDetails
     */
    public function __construct(
        public readonly string $amount,
        public readonly string $currency,
        public readonly string $periodUnit,
        public readonly string $periodCount,
        public readonly string $recipient = '',
        public readonly string $subscriptionExpires = '',
        public readonly string $description = '',
        public readonly string $externalId = '',
        public readonly ?array $methodDetails = null,
    ) {
        self::assertPositiveDecimal($amount, 'amount');
        if ($currency === '') {
            throw new InvalidArgumentException('currency is required');
        }
        self::normalizePeriodUnit($periodUnit);
        self::assertPositiveDecimal($periodCount, 'periodCount');
        if ($subscriptionExpires !== '') {
            self::parseSubscriptionExpires($subscriptionExpires);
        }
    }

    /**
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        $value = [
            'amount' => $this->amount,
            'currency' => $this->currency,
            'periodUnit' => $this->periodUnit,
            'periodCount' => $this->periodCount,
        ];
        if ($this->recipient !== '') {
            $value['recipient'] = $this->recipient;
        }
        if ($this->subscriptionExpires !== '') {
            $value['subscriptionExpires'] = $this->subscriptionExpires;
        }
        if ($this->description !== '') {
            $value['description'] = $this->description;
        }
        if ($this->externalId !== '') {
            $value['externalId'] = $this->externalId;
        }
        if ($this->methodDetails !== null) {
            $value['methodDetails'] = $this->methodDetails;
        }

        return $value;
    }

    /**
     * @param array<string, mixed> $value
     */
    public static function fromArray(array $value): self
    {
        $methodDetails = $value['methodDetails'] ?? null;
        if ($methodDetails !== null && !is_array($methodDetails)) {
            throw new InvalidArgumentException('methodDetails must be an object');
        }

        /** @var array<string, mixed>|null $methodDetails */
        return new self(
            amount: (string)($value['amount'] ?? ''),
            currency: (string)($value['currency'] ?? ''),
            periodUnit: (string)($value['periodUnit'] ?? ''),
            periodCount: (string)($value['periodCount'] ?? ''),
            recipient: (string)($value['recipient'] ?? ''),
            subscriptionExpires: (string)($value['subscriptionExpires'] ?? ''),
            description: (string)($value['description'] ?? ''),
            externalId: (string)($value['externalId'] ?? ''),
            methodDetails: $methodDetails,
        );
    }

    public static function normalizePeriodUnit(string $periodUnit): string
    {
        return match ($periodUnit) {
            self::PERIOD_DAY, self::PERIOD_WEEK, self::PERIOD_MONTH => $periodUnit,
            default => throw new InvalidArgumentException(sprintf('unsupported periodUnit: %s', $periodUnit)),
        };
    }

    public static function parseSubscriptionExpires(string $value): DateTimeImmutable
    {
        $parsed = DateTimeImmutable::createFromFormat(DATE_ATOM, $value);
        if ($parsed === false) {
            throw new InvalidArgumentException(sprintf('invalid subscriptionExpires: %s', $value));
        }

        return $parsed;
    }

    private static function assertPositiveDecimal(string $value, string $field): void
    {
        if ($value === '' || !ctype_digit($value)) {
            throw new InvalidArgumentException(sprintf('invalid %s: %s', $field, $value));
        }

        $canonical = ltrim($value, '0');
        if ($canonical === '' || $canonical !== $value) {
            throw new InvalidArgumentException(sprintf('invalid %s: %s', $field, $value));
        }
    }
}
